src/clients/Deluge.ts

```diff
@@ -0,0 +1,225 @@
+import { InjectionResult } from "../constants.js";
+import { CrossSeedError } from "../errors.js";
+import { Label, logger } from "../logger.js";
+import { Metafile } from "../parseTorrent.js";
+import { getRuntimeConfig } from "../runtimeConfig.js";
+import { Searchee } from "../searchee.js";
+import { TorrentClient } from "./TorrentClient.js";
+import { extractCredentialsFromUrl } from "../utils.js";
+import fetch, { Headers, Response } from "node-fetch";
+
+interface DelugeResponse {
+	error?: {
+		message?: string;
+		code?: DelugeErrorCode;
+	};
+	result?: string | boolean;
+}
+enum DelugeErrorCode {
+	NO_AUTH = 1,
+	BAD_METHOD = 2,
+	CALL_ERR = 3,
+	RPC_FAIL = 4,
+	BAD_JSON = 5,
+}
+interface TorrentInfo {
+	complete: boolean;
+	save_path?: string;
+}
+
+export default class Deluge implements TorrentClient {
+	private delugeCookie = null;
+	private delugeLabel = "cross-seed";
+	private isLabelEnabled: boolean;
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		await this.authenticate();
+		this.isLabelEnabled = await this.labelEnabled();
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<void> {
+		const { delugeRpcUrl } = getRuntimeConfig();
+		const { href, password } = extractCredentialsFromUrl(
+			delugeRpcUrl
+		).unwrapOrThrow(
+			new CrossSeedError("delugeRpcUrl must be percent-encoded")
+		);
+		if (!password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeRpcUrl. (e.g. http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const response = await this.call("auth.login", [password], 0);
+		if (!response.result) {
+			throw new CrossSeedError(
+				`reached deluge, but failed to authenticate: ${href}`
+			);
+		}
+	}
+
+	/**
+	 * ensures authentication and sends JSON-RPC calls to deluge
+	 */
+	private async call(method: string, params: object, retries = 1) {
+		const { delugeRpcUrl } = getRuntimeConfig();
+		const { href } = extractCredentialsFromUrl(delugeRpcUrl).unwrapOrThrow(
+			new CrossSeedError("delugeRpcUrl must be percent-encoded")
+		);
+		const headers = new Headers({ "Content-Type": "application/json" });
+		if (this.delugeCookie) headers.set("Cookie", this.delugeCookie);
+
+		let response: Response, json: DelugeResponse;
+		try {
+			response = await fetch(href, {
+				body: JSON.stringify({
+					...{ method: method, params: params },
+					id: Math.floor(Math.random() * 1000000),
+				}),
+				method: "POST",
+				headers: headers,
+			});
+			json = (await response.json()) as DelugeResponse;
+		} catch (NetworkError) {
+			throw new CrossSeedError(
+				`failed to establish a connection to deluge: ${href}`
+			);
+		}
+		if (json?.error?.code === DelugeErrorCode.NO_AUTH && retries > 0) {
+			this.delugeCookie = null;
+			await this.authenticate();
+			return this.call(method, params, 0);
+		}
```

Comment:
I'm not sure if I handled this properly, since it would reset everytime a new command is issued.

Are we supposed to CrossSeedError after retries fail?